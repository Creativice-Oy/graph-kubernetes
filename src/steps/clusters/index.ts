import {
  createDirectRelationship,
  createMappedRelationship,
  Entity,
  IntegrationStep,
  RelationshipClass,
  RelationshipDirection,
} from '@jupiterone/integration-sdk-core';
import { V1ConfigMap } from '@kubernetes/client-node';
import { IntegrationConfig, IntegrationStepContext } from '../../config';
import {
  CLUSTER_ENTITY_DATA_KEY,
  Entities,
  IntegrationSteps,
  Relationships,
} from '../constants';
import { createClusterEntity } from './converters';

export async function fetchClusterDetails(
  context: IntegrationStepContext,
): Promise<void> {
  const { instance, jobState } = context;
  const { name, id } = instance;

  const clusterEntity = createClusterEntity(name, id);

  await jobState.addEntity(clusterEntity);
  await jobState.setData(CLUSTER_ENTITY_DATA_KEY, clusterEntity);
}

export async function buildClusterResourcesRelationships(
  context: IntegrationStepContext,
): Promise<void> {
  const { jobState } = context;

  const clusterEntity = (await jobState.getData(
    CLUSTER_ENTITY_DATA_KEY,
  )) as Entity;

  await jobState.iterateEntities(
    {
      _type: Entities.NAMESPACE._type,
    },
    async (namespaceEntity) => {
      await jobState.addRelationship(
        createDirectRelationship({
          _class: RelationshipClass.CONTAINS,
          from: clusterEntity,
          to: namespaceEntity,
        }),
      );
    },
  );
}

function checkForAzureClusterAndReturnKey(data: Entity): null | string {
  if (!data._rawData || data._rawData?.length < 1) {
    return null;
  }

  // Lets turn this into V1ConfigMap as soon as possible so we have typings
  const configMap = data._rawData[0].rawData as V1ConfigMap;
  if (!configMap.metadata || !configMap.metadata.annotations) {
    return null;
  }

  const mustHaveProperties = [
    'subscriptions',
    'resourceGroups',
    'providers',
    'Microsoft.ContainerService',
    'managedClusters',
  ];

  for (const [_, value] of Object.entries(configMap.metadata.annotations)) {
    const annotationData = JSON.parse(value);
    const potentialKey: string = annotationData['data']['CLUSTER_RESOURCE_ID'];
    if (!potentialKey) {
      continue;
    }

    const keyPieces = potentialKey.split('/');
    if (mustHaveProperties.every((property) => keyPieces.includes(property))) {
      return potentialKey;
    }
  }

  return null;
}

export async function buildClusterCloudProviderRelationships(
  context: IntegrationStepContext,
): Promise<void> {
  const { jobState } = context;

  const clusterEntity = (await jobState.getData(
    CLUSTER_ENTITY_DATA_KEY,
  )) as Entity;

  await jobState.iterateEntities(
    {
      _type: Entities.CONFIGMAP._type,
    },
    async (configMapEntity) => {
      const azureKubernetesClusterKey = checkForAzureClusterAndReturnKey(
        configMapEntity,
      );
      if (azureKubernetesClusterKey) {
        await jobState.addRelationship(
          createMappedRelationship({
            _class: RelationshipClass.IS,
            _type: Relationships.CLUSTER_IS_AKS_CLUSTER._type,
            _mapping: {
              relationshipDirection: RelationshipDirection.FORWARD,
              sourceEntityKey: clusterEntity._key,
              targetFilterKeys: [['_type', '_key']],
              skipTargetCreation: true,
              targetEntity: {
                _type: Entities.CLUSTER._type,
                _key: azureKubernetesClusterKey,
              },
            },
          }),
        );
      }
    },
  );
}

export const clustersSteps: IntegrationStep<IntegrationConfig>[] = [
  {
    id: IntegrationSteps.CLUSTERS,
    name: 'Fetch Cluster',
    entities: [Entities.CLUSTER],
    relationships: [],
    dependsOn: [],
    executionHandler: fetchClusterDetails,
  },
  {
    id: IntegrationSteps.BUILD_CLUSTER_RESOURCES_RELATIONSHIPS,
    name: 'Build Cluster Resources Relationships',
    entities: [],
    relationships: [Relationships.CLUSTER_CONTAINS_NAMESPACE],
    dependsOn: [IntegrationSteps.CLUSTERS, IntegrationSteps.NAMESPACES],
    executionHandler: buildClusterResourcesRelationships,
  },
  {
    id: IntegrationSteps.BUILD_CLUSTER_CLOUD_PROVIDERS_RELATIONSHIPS,
    name: 'Build Cluster Cloud Providers Relationships',
    entities: [],
    relationships: [Relationships.CLUSTER_IS_AKS_CLUSTER],
    dependsOn: [IntegrationSteps.CLUSTERS, IntegrationSteps.CONFIGMAPS],
    executionHandler: buildClusterCloudProviderRelationships,
  },
];
