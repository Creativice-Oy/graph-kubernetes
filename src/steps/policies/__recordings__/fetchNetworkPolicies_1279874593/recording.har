{
  "log": {
    "_recordingName": "fetchNetworkPolicies",
    "creator": {
      "comment": "persister:JupiterOneIntegationFSPersister",
      "name": "Polly.JS",
      "version": "4.3.0"
    },
    "entries": [
      {
        "_id": "e262abbd5b6acc4c1139793afa880ed2",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 0,
          "cookies": [],
          "headers": [
            {
              "name": "accept",
              "value": "application/json"
            },
            {
              "name": "host",
              "value": "192.168.49.2:8443"
            }
          ],
          "headersSize": 121,
          "httpVersion": "HTTP/1.1",
          "method": "GET",
          "queryString": [
            {
              "name": "limit",
              "value": "250"
            }
          ],
          "url": "https://192.168.49.2:8443/api/v1/namespaces?limit=250"
        },
        "response": {
          "bodySize": 2056,
          "content": {
            "mimeType": "application/json",
            "size": 2056,
            "text": "{\"kind\":\"NamespaceList\",\"apiVersion\":\"v1\",\"metadata\":{\"resourceVersion\":\"1336630\"},\"items\":[{\"metadata\":{\"name\":\"default\",\"uid\":\"d5976ece-dc03-46a1-921f-cb6043e5f56e\",\"resourceVersion\":\"200\",\"creationTimestamp\":\"2021-04-27T20:57:19Z\",\"managedFields\":[{\"manager\":\"kube-apiserver\",\"operation\":\"Update\",\"apiVersion\":\"v1\",\"time\":\"2021-04-27T20:57:19Z\",\"fieldsType\":\"FieldsV1\",\"fieldsV1\":{\"f:status\":{\"f:phase\":{}}}}]},\"spec\":{\"finalizers\":[\"kubernetes\"]},\"status\":{\"phase\":\"Active\"}},{\"metadata\":{\"name\":\"kube-node-lease\",\"uid\":\"55ea52dd-382c-4b8c-8c24-bff59db4b817\",\"resourceVersion\":\"66\",\"creationTimestamp\":\"2021-04-27T20:57:17Z\",\"managedFields\":[{\"manager\":\"kube-apiserver\",\"operation\":\"Update\",\"apiVersion\":\"v1\",\"time\":\"2021-04-27T20:57:17Z\",\"fieldsType\":\"FieldsV1\",\"fieldsV1\":{\"f:status\":{\"f:phase\":{}}}}]},\"spec\":{\"finalizers\":[\"kubernetes\"]},\"status\":{\"phase\":\"Active\"}},{\"metadata\":{\"name\":\"kube-public\",\"uid\":\"35ca0e47-73b0-4e3d-9e2a-d8f3d6506d88\",\"resourceVersion\":\"58\",\"creationTimestamp\":\"2021-04-27T20:57:17Z\",\"managedFields\":[{\"manager\":\"kube-apiserver\",\"operation\":\"Update\",\"apiVersion\":\"v1\",\"time\":\"2021-04-27T20:57:17Z\",\"fieldsType\":\"FieldsV1\",\"fieldsV1\":{\"f:status\":{\"f:phase\":{}}}}]},\"spec\":{\"finalizers\":[\"kubernetes\"]},\"status\":{\"phase\":\"Active\"}},{\"metadata\":{\"name\":\"kube-system\",\"uid\":\"239ee6a3-23d6-4ca6-ae32-3b3fdda6a604\",\"resourceVersion\":\"34\",\"creationTimestamp\":\"2021-04-27T20:57:17Z\",\"managedFields\":[{\"manager\":\"kube-apiserver\",\"operation\":\"Update\",\"apiVersion\":\"v1\",\"time\":\"2021-04-27T20:57:17Z\",\"fieldsType\":\"FieldsV1\",\"fieldsV1\":{\"f:status\":{\"f:phase\":{}}}}]},\"spec\":{\"finalizers\":[\"kubernetes\"]},\"status\":{\"phase\":\"Active\"}},{\"metadata\":{\"name\":\"my-namespace\",\"uid\":\"0f9d0ac2-87db-41fe-ba0b-285778f29815\",\"resourceVersion\":\"1141617\",\"creationTimestamp\":\"2021-10-29T16:18:28Z\",\"managedFields\":[{\"manager\":\"kubectl-create\",\"operation\":\"Update\",\"apiVersion\":\"v1\",\"time\":\"2021-10-29T16:18:28Z\",\"fieldsType\":\"FieldsV1\",\"fieldsV1\":{\"f:status\":{\"f:phase\":{}}}}]},\"spec\":{\"finalizers\":[\"kubernetes\"]},\"status\":{\"phase\":\"Active\"}}]}"
          },
          "cookies": [],
          "headers": [
            {
              "name": "cache-control",
              "value": "no-cache, private"
            },
            {
              "name": "content-type",
              "value": "application/json"
            },
            {
              "name": "x-kubernetes-pf-flowschema-uid",
              "value": "210417ef-0504-45ee-ba4b-365d19207515"
            },
            {
              "name": "x-kubernetes-pf-prioritylevel-uid",
              "value": "f6528105-041a-490a-9dfb-a86e0c629d2a"
            },
            {
              "name": "date",
              "value": "Wed, 03 Nov 2021 18:26:47 GMT"
            },
            {
              "name": "connection",
              "value": "close"
            },
            {
              "name": "transfer-encoding",
              "value": "chunked"
            }
          ],
          "headersSize": 295,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2021-11-03T18:26:47.096Z",
        "time": 27,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 27
        }
      },
      {
        "_id": "f966932dcc2f7fa1fa3030bbee5bdf0d",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 0,
          "cookies": [],
          "headers": [
            {
              "name": "accept",
              "value": "application/json"
            },
            {
              "name": "host",
              "value": "192.168.49.2:8443"
            }
          ],
          "headersSize": 164,
          "httpVersion": "HTTP/1.1",
          "method": "GET",
          "queryString": [
            {
              "name": "limit",
              "value": "250"
            }
          ],
          "url": "https://192.168.49.2:8443/apis/networking.k8s.io/v1/namespaces/default/networkpolicies?limit=250"
        },
        "response": {
          "bodySize": 1826,
          "content": {
            "mimeType": "application/json",
            "size": 1826,
            "text": "{\"kind\":\"NetworkPolicyList\",\"apiVersion\":\"networking.k8s.io/v1\",\"metadata\":{\"resourceVersion\":\"1336630\"},\"items\":[{\"metadata\":{\"name\":\"test-network-policy\",\"namespace\":\"default\",\"uid\":\"9a6469a5-447a-4f0c-b012-c55988ce2f2f\",\"resourceVersion\":\"1267750\",\"generation\":1,\"creationTimestamp\":\"2021-11-02T15:28:06Z\",\"annotations\":{\"kubectl.kubernetes.io/last-applied-configuration\":\"{\\\"apiVersion\\\":\\\"networking.k8s.io/v1\\\",\\\"kind\\\":\\\"NetworkPolicy\\\",\\\"metadata\\\":{\\\"annotations\\\":{},\\\"name\\\":\\\"test-network-policy\\\",\\\"namespace\\\":\\\"default\\\"},\\\"spec\\\":{\\\"egress\\\":[{\\\"ports\\\":[{\\\"port\\\":5978,\\\"protocol\\\":\\\"TCP\\\"}],\\\"to\\\":[{\\\"ipBlock\\\":{\\\"cidr\\\":\\\"10.0.0.0/24\\\"}}]}],\\\"ingress\\\":[{\\\"from\\\":[{\\\"ipBlock\\\":{\\\"cidr\\\":\\\"172.17.0.0/16\\\",\\\"except\\\":[\\\"172.17.1.0/24\\\"]}},{\\\"namespaceSelector\\\":{\\\"matchLabels\\\":{\\\"project\\\":\\\"myproject\\\"}}},{\\\"podSelector\\\":{\\\"matchLabels\\\":{\\\"role\\\":\\\"frontend\\\"}}}],\\\"ports\\\":[{\\\"port\\\":6379,\\\"protocol\\\":\\\"TCP\\\"}]}],\\\"podSelector\\\":{\\\"matchLabels\\\":{\\\"role\\\":\\\"db\\\"}},\\\"policyTypes\\\":[\\\"Ingress\\\",\\\"Egress\\\"]}}\\n\"},\"managedFields\":[{\"manager\":\"kubectl-client-side-apply\",\"operation\":\"Update\",\"apiVersion\":\"networking.k8s.io/v1\",\"time\":\"2021-11-02T15:28:06Z\",\"fieldsType\":\"FieldsV1\",\"fieldsV1\":{\"f:metadata\":{\"f:annotations\":{\".\":{},\"f:kubectl.kubernetes.io/last-applied-configuration\":{}}},\"f:spec\":{\"f:egress\":{},\"f:ingress\":{},\"f:podSelector\":{},\"f:policyTypes\":{}}}}]},\"spec\":{\"podSelector\":{\"matchLabels\":{\"role\":\"db\"}},\"ingress\":[{\"ports\":[{\"protocol\":\"TCP\",\"port\":6379}],\"from\":[{\"ipBlock\":{\"cidr\":\"172.17.0.0/16\",\"except\":[\"172.17.1.0/24\"]}},{\"namespaceSelector\":{\"matchLabels\":{\"project\":\"myproject\"}}},{\"podSelector\":{\"matchLabels\":{\"role\":\"frontend\"}}}]}],\"egress\":[{\"ports\":[{\"protocol\":\"TCP\",\"port\":5978}],\"to\":[{\"ipBlock\":{\"cidr\":\"10.0.0.0/24\"}}]}],\"policyTypes\":[\"Ingress\",\"Egress\"]}}]}"
          },
          "cookies": [],
          "headers": [
            {
              "name": "cache-control",
              "value": "no-cache, private"
            },
            {
              "name": "content-type",
              "value": "application/json"
            },
            {
              "name": "x-kubernetes-pf-flowschema-uid",
              "value": "210417ef-0504-45ee-ba4b-365d19207515"
            },
            {
              "name": "x-kubernetes-pf-prioritylevel-uid",
              "value": "f6528105-041a-490a-9dfb-a86e0c629d2a"
            },
            {
              "name": "date",
              "value": "Wed, 03 Nov 2021 18:26:47 GMT"
            },
            {
              "name": "content-length",
              "value": "1826"
            },
            {
              "name": "connection",
              "value": "close"
            }
          ],
          "headersSize": 289,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2021-11-03T18:26:47.184Z",
        "time": 9,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 9
        }
      },
      {
        "_id": "386e311aa5778e493f3647082ba33d55",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 0,
          "cookies": [],
          "headers": [
            {
              "name": "accept",
              "value": "application/json"
            },
            {
              "name": "host",
              "value": "192.168.49.2:8443"
            }
          ],
          "headersSize": 172,
          "httpVersion": "HTTP/1.1",
          "method": "GET",
          "queryString": [
            {
              "name": "limit",
              "value": "250"
            }
          ],
          "url": "https://192.168.49.2:8443/apis/networking.k8s.io/v1/namespaces/kube-node-lease/networkpolicies?limit=250"
        },
        "response": {
          "bodySize": 117,
          "content": {
            "mimeType": "application/json",
            "size": 117,
            "text": "{\"kind\":\"NetworkPolicyList\",\"apiVersion\":\"networking.k8s.io/v1\",\"metadata\":{\"resourceVersion\":\"1336630\"},\"items\":[]}"
          },
          "cookies": [],
          "headers": [
            {
              "name": "cache-control",
              "value": "no-cache, private"
            },
            {
              "name": "content-type",
              "value": "application/json"
            },
            {
              "name": "x-kubernetes-pf-flowschema-uid",
              "value": "210417ef-0504-45ee-ba4b-365d19207515"
            },
            {
              "name": "x-kubernetes-pf-prioritylevel-uid",
              "value": "f6528105-041a-490a-9dfb-a86e0c629d2a"
            },
            {
              "name": "date",
              "value": "Wed, 03 Nov 2021 18:26:47 GMT"
            },
            {
              "name": "content-length",
              "value": "117"
            },
            {
              "name": "connection",
              "value": "close"
            }
          ],
          "headersSize": 288,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2021-11-03T18:26:47.201Z",
        "time": 6,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 6
        }
      },
      {
        "_id": "e6fbbd1e40374081e9b4bcaa02b4e8dd",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 0,
          "cookies": [],
          "headers": [
            {
              "name": "accept",
              "value": "application/json"
            },
            {
              "name": "host",
              "value": "192.168.49.2:8443"
            }
          ],
          "headersSize": 168,
          "httpVersion": "HTTP/1.1",
          "method": "GET",
          "queryString": [
            {
              "name": "limit",
              "value": "250"
            }
          ],
          "url": "https://192.168.49.2:8443/apis/networking.k8s.io/v1/namespaces/kube-public/networkpolicies?limit=250"
        },
        "response": {
          "bodySize": 117,
          "content": {
            "mimeType": "application/json",
            "size": 117,
            "text": "{\"kind\":\"NetworkPolicyList\",\"apiVersion\":\"networking.k8s.io/v1\",\"metadata\":{\"resourceVersion\":\"1336630\"},\"items\":[]}"
          },
          "cookies": [],
          "headers": [
            {
              "name": "cache-control",
              "value": "no-cache, private"
            },
            {
              "name": "content-type",
              "value": "application/json"
            },
            {
              "name": "x-kubernetes-pf-flowschema-uid",
              "value": "210417ef-0504-45ee-ba4b-365d19207515"
            },
            {
              "name": "x-kubernetes-pf-prioritylevel-uid",
              "value": "f6528105-041a-490a-9dfb-a86e0c629d2a"
            },
            {
              "name": "date",
              "value": "Wed, 03 Nov 2021 18:26:47 GMT"
            },
            {
              "name": "content-length",
              "value": "117"
            },
            {
              "name": "connection",
              "value": "close"
            }
          ],
          "headersSize": 288,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2021-11-03T18:26:47.213Z",
        "time": 9,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 9
        }
      },
      {
        "_id": "f4d19e8f5d15c7c0decb22ffe99ea768",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 0,
          "cookies": [],
          "headers": [
            {
              "name": "accept",
              "value": "application/json"
            },
            {
              "name": "host",
              "value": "192.168.49.2:8443"
            }
          ],
          "headersSize": 168,
          "httpVersion": "HTTP/1.1",
          "method": "GET",
          "queryString": [
            {
              "name": "limit",
              "value": "250"
            }
          ],
          "url": "https://192.168.49.2:8443/apis/networking.k8s.io/v1/namespaces/kube-system/networkpolicies?limit=250"
        },
        "response": {
          "bodySize": 117,
          "content": {
            "mimeType": "application/json",
            "size": 117,
            "text": "{\"kind\":\"NetworkPolicyList\",\"apiVersion\":\"networking.k8s.io/v1\",\"metadata\":{\"resourceVersion\":\"1336630\"},\"items\":[]}"
          },
          "cookies": [],
          "headers": [
            {
              "name": "cache-control",
              "value": "no-cache, private"
            },
            {
              "name": "content-type",
              "value": "application/json"
            },
            {
              "name": "x-kubernetes-pf-flowschema-uid",
              "value": "210417ef-0504-45ee-ba4b-365d19207515"
            },
            {
              "name": "x-kubernetes-pf-prioritylevel-uid",
              "value": "f6528105-041a-490a-9dfb-a86e0c629d2a"
            },
            {
              "name": "date",
              "value": "Wed, 03 Nov 2021 18:26:47 GMT"
            },
            {
              "name": "content-length",
              "value": "117"
            },
            {
              "name": "connection",
              "value": "close"
            }
          ],
          "headersSize": 288,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2021-11-03T18:26:47.225Z",
        "time": 7,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 7
        }
      },
      {
        "_id": "e1545f8ec25ae97381856d30398e9706",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 0,
          "cookies": [],
          "headers": [
            {
              "name": "accept",
              "value": "application/json"
            },
            {
              "name": "host",
              "value": "192.168.49.2:8443"
            }
          ],
          "headersSize": 169,
          "httpVersion": "HTTP/1.1",
          "method": "GET",
          "queryString": [
            {
              "name": "limit",
              "value": "250"
            }
          ],
          "url": "https://192.168.49.2:8443/apis/networking.k8s.io/v1/namespaces/my-namespace/networkpolicies?limit=250"
        },
        "response": {
          "bodySize": 117,
          "content": {
            "mimeType": "application/json",
            "size": 117,
            "text": "{\"kind\":\"NetworkPolicyList\",\"apiVersion\":\"networking.k8s.io/v1\",\"metadata\":{\"resourceVersion\":\"1336630\"},\"items\":[]}"
          },
          "cookies": [],
          "headers": [
            {
              "name": "cache-control",
              "value": "no-cache, private"
            },
            {
              "name": "content-type",
              "value": "application/json"
            },
            {
              "name": "x-kubernetes-pf-flowschema-uid",
              "value": "210417ef-0504-45ee-ba4b-365d19207515"
            },
            {
              "name": "x-kubernetes-pf-prioritylevel-uid",
              "value": "f6528105-041a-490a-9dfb-a86e0c629d2a"
            },
            {
              "name": "date",
              "value": "Wed, 03 Nov 2021 18:26:47 GMT"
            },
            {
              "name": "content-length",
              "value": "117"
            },
            {
              "name": "connection",
              "value": "close"
            }
          ],
          "headersSize": 288,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2021-11-03T18:26:47.236Z",
        "time": 9,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 9
        }
      }
    ],
    "pages": [],
    "version": "1.2"
  }
}
