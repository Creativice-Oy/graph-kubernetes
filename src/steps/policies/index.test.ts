import { fetchPodSecurityPolicies } from '.';
import { createDataCollectionTest } from '../../../test/recording';
import { integrationConfig } from '../../../test/config';
import { Entities, Relationships } from '../constants';
import { fetchClusterDetails } from '../clusters';
import { RelationshipClass } from '@jupiterone/data-model';

describe('#fetchPodSecurityPolicies', () => {
  test('should collect data', async () => {
    await createDataCollectionTest({
      recordingName: 'fetchPodSecurityPolicies',
      recordingDirectory: __dirname,
      integrationConfig,
      stepFunctions: [fetchClusterDetails, fetchPodSecurityPolicies],
      entitySchemaMatchers: [
        {
          _type: Entities.POD_SECURITY_POLICY._type,
          matcher: {
            _class: ['Configuration'],
            schema: {
              additionalProperties: false,
              properties: {
                _type: { const: 'kube_pod_security_policy' },
                _rawData: {
                  type: 'array',
                  items: { type: 'object' },
                },
                name: { type: 'string' },
                displayName: { type: 'string' },
                generation: { type: 'number' },
                deletionGracePeriodSeconds: { type: 'string' },
                resourceVersion: { type: 'string' },
                createdOn: { type: 'number' },

                allowPrivilegeEscalation: { type: 'boolean' },
                allowedCapabilities: {
                  type: 'array',
                  items: { type: 'string' },
                },
                allowedProcMountTypes: {
                  type: 'array',
                  items: { type: 'string' },
                },
                allowedUnsafeSysctls: {
                  type: 'array',
                  items: { type: 'string' },
                },
                defaultAddCapabilities: {
                  type: 'array',
                  items: { type: 'string' },
                },
                defaultAllowPrivilegeEscalation: { type: 'boolean' },
                forbiddenSysctls: { type: 'array', items: { type: 'string' } },
                hostIPC: { type: 'boolean' },
                hostNetwork: { type: 'boolean' },
                hostPID: { type: 'boolean' },
                privileged: { type: 'boolean' },
                readOnlyRootFilesystem: { type: 'boolean' },
                requiredDropCapabilities: {
                  type: 'array',
                  items: { type: 'string' },
                },
                'runAsGroup.range': {
                  type: 'array',
                  items: { type: 'string' },
                },
                'runAsGroup.rule': { type: 'string' },
                'runAsUser.range': { type: 'array', items: { type: 'string' } },
                'runAsUser.rule': { type: 'string' },
                'runtimeClass.allowedRuntimeClassNames': {
                  type: 'array',
                  items: { type: 'string' },
                },
                'runtimeClass.defaultRuntimeClassName': { type: 'string' },
                'seLinux.rule': { type: 'string' },
                'seLinux.seLinuxOptions.level': { type: 'string' },
                'seLinux.seLinuxOptions.role': { type: 'string' },
                'seLinux.seLinuxOptions.type': { type: 'string' },
                'seLinux.seLinuxOptions.user': { type: 'string' },
                'supplementalGroups.range': {
                  type: 'array',
                  items: { type: 'string' },
                },
                'supplementalGroups.rule': { type: 'string' },
              },
            },
          },
        },
      ],
      relationshipSchemaMatchers: [
        {
          _type: Relationships.CLUSTER_CONTAINS_POD_SECURITY_POLICY._type,
          matcher: {
            schema: {
              properties: {
                _class: { const: RelationshipClass.CONTAINS },
                _type: { const: 'kube_cluster_contains_pod_security_policy' },
              },
            },
          },
        },
      ],
    });
  });
});
