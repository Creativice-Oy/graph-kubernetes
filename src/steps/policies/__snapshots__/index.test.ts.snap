// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#fetchPodSecurityPolicies should collect data: jobState 1`] = `
Object {
  "collectedEntities": Array [
    Object {
      "_class": Array [
        "Cluster",
      ],
      "_key": "kube_cluster:local-integration-instance",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "name": "Local Integration",
          },
        },
      ],
      "_type": "kube_cluster",
      "createdOn": undefined,
      "displayName": "Local Integration",
      "name": "Local Integration",
    },
    Object {
      "_class": Array [
        "Configuration",
      ],
      "_key": "d2ddbbb0-5714-4597-bc8b-cc521a978870",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": V1beta1PodSecurityPolicy {
            "apiVersion": undefined,
            "kind": undefined,
            "metadata": V1ObjectMeta {
              "annotations": Object {
                "kubectl.kubernetes.io/last-applied-configuration": "{\\"apiVersion\\":\\"policy/v1beta1\\",\\"kind\\":\\"PodSecurityPolicy\\",\\"metadata\\":{\\"annotations\\":{},\\"name\\":\\"example\\"},\\"spec\\":{\\"fsGroup\\":{\\"rule\\":\\"RunAsAny\\"},\\"privileged\\":false,\\"runAsUser\\":{\\"rule\\":\\"RunAsAny\\"},\\"seLinux\\":{\\"rule\\":\\"RunAsAny\\"},\\"supplementalGroups\\":{\\"ranges\\":[{\\"max\\":65535,\\"min\\":1}],\\"rule\\":\\"MustRunAs\\"},\\"volumes\\":[\\"*\\"]}}
",
              },
              "clusterName": undefined,
              "creationTimestamp": 2021-11-01T17:48:24.000Z,
              "deletionGracePeriodSeconds": undefined,
              "deletionTimestamp": undefined,
              "finalizers": undefined,
              "generateName": undefined,
              "generation": undefined,
              "labels": undefined,
              "managedFields": Array [
                V1ManagedFieldsEntry {
                  "apiVersion": "policy/v1beta1",
                  "fieldsType": "FieldsV1",
                  "fieldsV1": Object {
                    "f:metadata": Object {
                      "f:annotations": Object {
                        ".": Object {},
                        "f:kubectl.kubernetes.io/last-applied-configuration": Object {},
                      },
                    },
                    "f:spec": Object {
                      "f:allowPrivilegeEscalation": Object {},
                      "f:fsGroup": Object {
                        "f:rule": Object {},
                      },
                      "f:runAsUser": Object {
                        "f:rule": Object {},
                      },
                      "f:seLinux": Object {
                        "f:rule": Object {},
                      },
                      "f:supplementalGroups": Object {
                        "f:ranges": Object {},
                        "f:rule": Object {},
                      },
                      "f:volumes": Object {},
                    },
                  },
                  "manager": "kubectl-client-side-apply",
                  "operation": "Update",
                  "time": 2021-11-01T18:27:53.000Z,
                },
              ],
              "name": "example",
              "namespace": undefined,
              "ownerReferences": undefined,
              "resourceVersion": "1214124",
              "selfLink": undefined,
              "uid": "d2ddbbb0-5714-4597-bc8b-cc521a978870",
            },
            "spec": V1beta1PodSecurityPolicySpec {
              "allowPrivilegeEscalation": true,
              "allowedCSIDrivers": undefined,
              "allowedCapabilities": undefined,
              "allowedFlexVolumes": undefined,
              "allowedHostPaths": undefined,
              "allowedProcMountTypes": undefined,
              "allowedUnsafeSysctls": undefined,
              "defaultAddCapabilities": undefined,
              "defaultAllowPrivilegeEscalation": undefined,
              "forbiddenSysctls": undefined,
              "fsGroup": V1beta1FSGroupStrategyOptions {
                "ranges": undefined,
                "rule": "RunAsAny",
              },
              "hostIPC": undefined,
              "hostNetwork": undefined,
              "hostPID": undefined,
              "hostPorts": undefined,
              "privileged": undefined,
              "readOnlyRootFilesystem": undefined,
              "requiredDropCapabilities": undefined,
              "runAsGroup": undefined,
              "runAsUser": V1beta1RunAsUserStrategyOptions {
                "ranges": undefined,
                "rule": "RunAsAny",
              },
              "runtimeClass": undefined,
              "seLinux": V1beta1SELinuxStrategyOptions {
                "rule": "RunAsAny",
                "seLinuxOptions": undefined,
              },
              "supplementalGroups": V1beta1SupplementalGroupsStrategyOptions {
                "ranges": Array [
                  V1beta1IDRange {
                    "max": 65535,
                    "min": 1,
                  },
                ],
                "rule": "MustRunAs",
              },
              "volumes": Array [
                "*",
              ],
            },
          },
        },
      ],
      "_type": "kube_pod_security_policy",
      "allowPrivilegeEscalation": true,
      "allowedCapabilities": undefined,
      "allowedProcMountTypes": undefined,
      "allowedUnsafeSysctls": undefined,
      "createdOn": 1635788904000,
      "defaultAddCapabilities": undefined,
      "defaultAllowPrivilegeEscalation": undefined,
      "deletionGracePeriodSeconds": undefined,
      "displayName": "example",
      "forbiddenSysctls": undefined,
      "generation": undefined,
      "hostIPC": false,
      "hostNetwork": false,
      "hostPID": false,
      "name": "example",
      "privileged": false,
      "readOnlyRootFilesystem": false,
      "requiredDropCapabilities": undefined,
      "resourceVersion": "1214124",
      "runAsGroup.range": undefined,
      "runAsGroup.rule": undefined,
      "runAsUser.range": undefined,
      "runAsUser.rule": "RunAsAny",
      "runtimeClass.allowedRuntimeClassNames": undefined,
      "runtimeClass.defaultRuntimeClassName": undefined,
      "seLinux.rule": "RunAsAny",
      "seLinux.seLinuxOptions.level": undefined,
      "seLinux.seLinuxOptions.role": undefined,
      "seLinux.seLinuxOptions.type": undefined,
      "seLinux.seLinuxOptions.user": undefined,
      "supplementalGroups.range": Array [
        "1, 65535",
      ],
      "supplementalGroups.rule": "MustRunAs",
    },
  ],
  "collectedRelationships": Array [
    Object {
      "_class": "CONTAINS",
      "_fromEntityKey": "kube_cluster:local-integration-instance",
      "_key": "kube_cluster:local-integration-instance|contains|d2ddbbb0-5714-4597-bc8b-cc521a978870",
      "_toEntityKey": "d2ddbbb0-5714-4597-bc8b-cc521a978870",
      "_type": "kube_cluster_contains_pod_security_policy",
      "displayName": "CONTAINS",
    },
  ],
  "encounteredTypes": Array [
    "kube_cluster",
    "kube_pod_security_policy",
    "kube_cluster_contains_pod_security_policy",
  ],
  "numCollectedEntities": 2,
  "numCollectedRelationships": 1,
}
`;
